system-helper/flatpak-system-helper.c

```diff
@@ -51,6 +59,239 @@ G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitAuthorizationResult, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitDetails, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitSubject, g_object_unref)
 
+typedef struct
+{
+  FlatpakSystemHelper *object;
+  GDBusMethodInvocation *invocation;
+  GCancellable *cancellable;
+
+  guint watch_id;
+  uid_t uid;
+
+  gint ref_count;
+
+  gchar *ref;
+  gchar *commit;
+  gchar *root_dir_name;
+  gchar *user_dir_name;
+  gchar *child_repo_path;
+
+  GUnixMountMonitor *mount_monitor;
+  gulong mount_monitor_id;
+  guint unsuccessful_mount_timeout_id;
+  GSubprocess *bindfs;
+} OngoingPull;
+
+static void
+ongoing_pull_free (OngoingPull *pull)
+{
+  g_autoptr(GFile) root_dir = NULL;
+  g_autoptr(GFile) user_dir = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  root_dir = g_file_new_for_path (pull->root_dir_name);
+  user_dir = g_file_new_for_path (pull->user_dir_name);
+
+  g_cancellable_cancel (pull->cancellable);
+
+  g_clear_object (&pull->bindfs);
+
+  if (!flatpak_rm_rf (root_dir, NULL, &local_error))
```

Comment:
A question about the partial pull case: Won't this rm -rf prevent us from ever finding a partial pull ?

Reply:
Well, there are two scenario's here:
* error and client calls CancelPull - In that case, the partial pull will be cleared
* other cases like abrupt system shutdown - maybe that's when it's useful. 
(certainly this is a `needs improvement` area)

Reply:
I was wondering about a case where a client starts a pull, and then crashes (from the system helper perspective, it'll just drop off the bus)